### PR TITLE
feat: Support `logprobs` for vLLM models in OpenAI Frontend

### PR DESCRIPTION
**Pull request by pskiran1**
#### What does the PR do?

This PR adds support for `logprobs` functionality in the OpenAI-compatible frontend for vLLM models. The feature allows users to request detailed probability information for generated tokens, which is useful for understanding model confidence and exploring alternative completions.

**Key changes:**

- Added logprobs support for both chat completions and standard completions endpoints
- Implemented conversion from vLLM's logprobs format to OpenAI's format
- Added comprehensive test coverage for logprobs functionality, including validation and streaming

File | Description
-- | --
python/openai/openai_frontend/engine/utils/triton.py | Added helper functions to parse and convert logprobs from vLLM responses to OpenAI format for both chat and completion endpoints
python/openai/openai_frontend/engine/triton_engine.py | Integrated logprobs support into request handling, validation, and response generation for both streaming and non-streaming modes
python/openai/tests/test_openai_client.py | Added async tests for logprobs functionality using the OpenAI client library, including validation tests
python/openai/tests/test_chat_completions.py | Added HTTP-level tests for chat completions with logprobs, including edge cases and validation
python/openai/tests/test_completions.py | Added HTTP-level tests for completions with logprobs, including edge cases and validation

**Background:**

https://platform.openai.com/docs/api-reference/completions/create
https://platform.openai.com/docs/api-reference/chat/create

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Test plan: 

- CI Pipeline ID: 39072711

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: #xxx

### CODE REVIEW

File: python/openai/openai_frontend/engine/triton_engine.py [hunk 1]
```diff
@@ -790,8 +818,19 @@ def _validate_chat_request(
...
             )
 
-        if request.logit_bias is not None or request.logprobs:
-            raise ClientError("logit bias and log probs not currently supported")
+        if request.logit_bias is not None:
+            raise ClientError("logit bias is not currently supported")
+
+        # Logprobs are only supported for vLLM backend currently
+        if metadata.backend != "vllm" and (
+            request.logprobs is not None or request.top_logprobs is not None
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply python/openai/openai_frontend/engine/triton_engine.py:827]: Default value for `request.logprobs` and `request.top_logprobs` is `None` in the schema file.

File: python/openai/openai_frontend/engine/triton_engine.py [hunk 2]
```diff
@@ -790,8 +818,19 @@ def _validate_chat_request(
...
+        ):
+            raise ClientError(
+                "logprobs are currently available only for the vLLM backend"
+            )
+
+        if request.top_logprobs and not request.logprobs:
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply python/openai/openai_frontend/engine/triton_engine.py]: Thanks.

File: python/openai/openai_frontend/engine/utils/triton.py [hunk 3]
```diff
@@ -102,11 +109,31 @@ def _create_vllm_generate_request(
         # If neither is set, use a default value for max_tokens
         else:
             sampling_parameters["max_tokens"] = default_max_tokens
+
+        # Handle logprobs for chat completions
+        # OpenAI API: logprobs (bool), top_logprobs (int 0-20)
+        # vLLM API: logprobs (int) - number of top token logprobs to return
+        if request.logprobs and request.top_logprobs is not None:
```
Comment (whoisj) [inline python/openai/openai_frontend/engine/utils/triton.py:116]: are we guaranteed `request.logprobs` exists? if not, what happens when the property doesn't even exist?

File: python/openai/openai_frontend/engine/utils/triton.py [hunk 4]
```diff
@@ -378,6 +405,167 @@ def _get_output(response: tritonserver._api._response.InferenceResponse) -> str:
...
+    for token_logprobs_dict in vllm_logprobs:
+        if not token_logprobs_dict:
+            continue
+
+        # Sort by rank to identify the selected token (rank=1 is always the chosen token)
+        sorted_tokens = sorted(
+            token_logprobs_dict.items(), key=lambda x: x[1].get("rank", sys.maxsize)
+        )
+
+        if not sorted_tokens:
```
Comment (whoisj) [inline python/openai/openai_frontend/engine/utils/triton.py]: what would `not sorted_tokens` mean here?

File: python/openai/openai_frontend/engine/utils/triton.py (hunk 3 above)
Comment (pskiran1) [reply python/openai/openai_frontend/engine/utils/triton.py:116]: Yes, it will always be present because we defined the schema in `python/openai/openai_frontend/schemas/openai.py`.

File: python/openai/openai_frontend/engine/utils/triton.py (hunk 4 above)
Comment (pskiran1) [reply python/openai/openai_frontend/engine/utils/triton.py]: This check was intended to verify if `sorted_tokens` is an empty dict. However, this check is unnecessary because earlier code guarantees it contains elements, so it has been removed. Thank you.

File: python/openai/tests/test_completions.py [hunk 5]
```diff
@@ -397,3 +397,96 @@ def test_usage_response(self, client, model: str, prompt: str):
...
+                assert isinstance(logprobs["token_logprobs"][i], (int, float))
+                assert isinstance(logprobs["text_offset"][i], int)
+                assert isinstance(logprobs["top_logprobs"][i], dict)
+
+                # Validate top_logprobs dict contains token -> logprob mappings
+                for token, logprob in logprobs["top_logprobs"][i].items():
+                    assert isinstance(token, str)
+                    assert isinstance(logprob, (int, float))
+
+        # Test that logprobs=0 returns no logprobs
```
Comment (yinggeh) [inline python/openai/tests/test_completions.py]: Please put logprobs=0 and invalid logprobs in two separate tests.

File: python/openai/tests/test_chat_completions.py [hunk 6]
```diff
@@ -548,6 +541,118 @@ def test_usage_response(self, client, model: str, messages: List[dict]):
...
+                assert isinstance(token_logprob["bytes"], list)
+                assert isinstance(token_logprob["top_logprobs"], list)
+
+                # Validate top_logprobs structure
+                for top_logprob in token_logprob["top_logprobs"]:
+                    assert "token" in top_logprob
+                    assert "logprob" in top_logprob
+                    assert "bytes" in top_logprob
+
+        # Test that logprobs=False returns no logprobs
```
Comment (yinggeh) [inline python/openai/tests/test_chat_completions.py]: Please put logprobs=False and invalid logprobs in two separate tests. Use `@pytest.mark.parametrize` for multiiple inputs.

File: python/openai/tests/test_openai_client.py [hunk 7]
```diff
@@ -482,3 +482,228 @@ async def test_stream_options_without_streaming(
...
+                    logprobs=True,
+                    top_logprobs=2,
+                    max_tokens=10,
+                )
+            assert "logprobs are currently available only for the vLLM backend" in str(
+                exc_info.value
+            )
+            return
+
+        chat_completion = await client.chat.completions.create(
```
Comment (yinggeh) [inline python/openai/tests/test_openai_client.py:510]: Is it better to send non-stream and steam requests and compare the output values to be the same, similar to `test_chat_streaming`?

File: python/openai/openai_frontend/engine/triton_engine.py [hunk 8]
```diff
@@ -605,6 +624,15 @@ async def _streaming_chat_iterator(
...
             previous_text = current_text
 
+            # Parse logprobs for this chunk if requested
+            chunk_logprobs = None
+            if request.logprobs:
+                openai_logprobs = _get_openai_chat_format_logprobs_from_vllm_response(
+                    response
+                )
+                if openai_logprobs:
+                    chunk_logprobs = Logprobs2(content=openai_logprobs)
```
Comment (yinggeh) [inline python/openai/openai_frontend/engine/triton_engine.py]: Can you rename Logprobs2 to something more meaningful?

File: python/openai/openai_frontend/engine/triton_engine.py [hunk 9]
```diff
@@ -846,16 +885,32 @@ async def _streaming_completion_iterator(
         model = request.model
         include_usage = request.stream_options and request.stream_options.include_usage
         usage_accumulator = _StreamingUsageAccumulator(backend)
+        current_offset = 0
```
Comment (yinggeh) [inline python/openai/openai_frontend/engine/triton_engine.py:889]: Is this parameter necessary? Should it be a local variable?

File: python/openai/tests/test_openai_client.py (hunk 7 above)
Comment (pskiran1) [reply python/openai/tests/test_openai_client.py:510]: Added changes to compare the counts of streaming and non-streaming logprobs and tokens. Unable to compare logprobs values due to floating point discrepancies exceeding 1e-2 for some values.
Ex: `-0.11291083693504333` vs `-0.12702862918376923`, `-0.002805228577926755` vs `-0.0024760086089372635` and `-0.1270265281200409` vs `-0.1270267367362976`

File: python/openai/openai_frontend/engine/triton_engine.py (hunk 9 above)
Comment (pskiran1) [reply python/openai/openai_frontend/engine/triton_engine.py:889]: `current_offset` is a local variable (not a parameter) and is necessary to accumulate text position across streaming chunks, enabling correct adjustment of `text_offset` values in logprobs.

Comment (yinggeh) [reply python/openai/openai_frontend/engine/triton_engine.py:889]: My bad. Misread the code

File: python/openai/tests/test_openai_client.py (hunk 7 above)
Comment (yinggeh) [reply python/openai/tests/test_openai_client.py:510]: You can use `np.allclose`. See examples in [test_embeddings.py](https://github.com/triton-inference-server/server/blob/main/python/openai/tests/test_embeddings.py)

Comment (pskiran1) [reply python/openai/tests/test_openai_client.py:510]: Modified to use `np.allclose`. Thank you.
